tools/rustdoc/rustdoc_test_writer.rs

```diff
@@ -0,0 +1,211 @@
+}
+
+/// Parse command line arguments
+fn parse_args() -> Options {
+    let args: Vec<String> = env::args().into_iter().collect();
+    let (writer_args, action_args) = {
+        let split = args
+            .iter()
+            .position(|arg| arg == "--")
+            .expect("Unable to find split identifier `--`");
+
+        // Converting each set into a vector makes them easier to parse in
+        // the absence of nightly features
+        let (writer, action) = args.split_at(split);
+        (writer.to_vec(), action.to_vec())
+    };
+
+    // Remove the leading `--` which is expected to be the first
+    // item in `action_args`
+    debug_assert_eq!(action_args[0], "--");
+    let action_argv = action_args[1..].to_vec();
+
+    let output = writer_args
+        .iter()
+        .find(|arg| arg.starts_with("--output="))
+        .and_then(|arg| arg.splitn(2, '=').last())
+        .map(PathBuf::from)
+        .expect("Missing `--output` argument");
+
+    let (strip_substring_args, writer_args): (Vec<String>, Vec<String>) = writer_args
+        .into_iter()
+        .partition(|arg| arg.starts_with("--strip_substring="));
+
+    let mut strip_substrings: Vec<String> = strip_substring_args
+        .into_iter()
+        .map(|arg| {
+            arg.splitn(2, '=')
+                .last()
+                .expect("--strip_substring arguments must have assignments using `=`")
+                .to_owned()
+        })
+        .collect();
+
+    // Strip substrings should always be in reverse order of the length of each
+    // string so when filtering we know that the longer strings are checked
+    // first in order to avoid cases where shorter strings might match longer ones.
+    strip_substrings.sort_by_key(|b| Reverse(b.len()));
+    strip_substrings.dedup();
+
+    let mut env_keys: Vec<String> = writer_args
+        .into_iter()
+        .filter(|arg| arg.starts_with("--action_env="))
+        .map(|arg| {
+            arg.splitn(2, '=')
+                .last()
+                .expect("--env arguments must have assignments using `=`")
+                .to_owned()
+        })
+        .collect();
+    env_keys.sort();
+    env_keys.dedup();
+
+    Options {
+        env_keys,
+        strip_substrings,
+        output,
+        action_argv,
+    }
+}
+
+/// Write a unix compatible test runner
+fn write_test_runner_unix(
+    path: &Path,
+    env: &HashMap<String, String>,
+    argv: &[String],
+    strip_substrings: &[String],
+) {
+    let mut content = vec![
+        "#!/usr/bin/env bash".to_owned(),
+        "".to_owned(),
+        "exec env - \\".to_owned(),
+    ];
+
+    content.extend(env.iter().map(|(key, val)| format!("{}='{}' \\", key, val)));
+
+    let argv_str = argv
+        .iter()
+        // Remove any substrings found in the argument
+        .map(|arg| {
+            match strip_substrings
+                .iter()
+                .find(|substring| arg.contains(substring.as_str()))
```

Comment:
It's intentional that this looks for one match then continues. Since the substrings are reverse sorted by the length of the string, it means the first match should be the correct match for this particular argument. I was worried about shorter strings getting in and there being an incorrect replacement. Though, I don't think there's much risk in a double replacement in how this turned out. Do you still think I should be trying to do replacements on the entire list?

Reply:
I think it's fine if we document that the idea is these are non-overlapping paths to roots, at at most one of them is expected to match - right now that assumption is a bit scattered around the code... But just replacing all of them would avoid needing to document that assumption at all :)

Reply:
Updated, now all substrings are replaced 😄